compiler/rustc_session/src/config.rs

```diff
@@ -2678,7 +2701,12 @@ pub fn parse_crate_types_from_list(list_list: Vec<String>) -> Result<Vec<CrateTy
                 "cdylib" => CrateType::Cdylib,
                 "bin" => CrateType::Executable,
                 "proc-macro" => CrateType::ProcMacro,
-                _ => return Err(format!("unknown crate type: `{part}`")),
+                _ => {
+                    return Err(format!(
+                        "unknown crate type `{part}`, expected one of: {display}",
+                        display = CrateType::shorthands_display()
+                    ));
+                }
```

Comment:
Suggestion: as @ollie27 mentioned above, we forgot about the "lib" being a valid accepted value which is not part of the `CrateType` enum ("lib" currently defaults to `default_lib_output()`).

Instead of adding the `shorthand` and `shorthands_display` methods to `CrateType`, could you instead slightly refactor this `match part` logic, to something like

```rs
const SUPPORTED_CRATE_TYPES: &str = ["lib", /* .. accepted values from the match arms */];

if SUPPORTED_CRATE_TYPES.contains(&part) {
    // blah
} else {
    return Err(format!("unknown crate type: `{}`, expected one of: ..."));
}
```

And could you add the colon back after "unknown crate type" (sorry)?

Reply:
I think we can keep the error in the `_` match arm, because we still need to match the input to enum variant.
So maybe something like this?
```rs
pub fn parse_crate_types_from_list(list_list: Vec<String>) -> Result<Vec<CrateType>, String> {
    const SUPPORTED_CRATE_TYPES: &[&str] = &[
        "lib",
        "rlib",
        "staticlib",
        "dylib", 
        "cdylib",
        "bin",
        "proc-macro",
    ];

    let mut crate_types: Vec<CrateType> = Vec::new();
    for unparsed_crate_type in &list_list {
        for part in unparsed_crate_type.split(',') {
            let new_part = match part {
                "lib" => default_lib_output(),
                "rlib" => CrateType::Rlib,
                "staticlib" => CrateType::Staticlib,
                "dylib" => CrateType::Dylib,
                "cdylib" => CrateType::Cdylib,
                "bin" => CrateType::Executable,
                "proc-macro" => CrateType::ProcMacro,
                _ => return Err(format!(
                    "unknown crate type: `{part}`, expected one of: {display}",
                    display = SUPPORTED_CRATE_TYPES.join(", ")
                )),
            };

            if !crate_types.contains(&new_part) {
                crate_types.push(new_part)
            }
        }
    }

    Ok(crate_types)
}
```

Possibly the `SUPPORTED_CRATE_TYPES` could be inlined into `format!`

```diff
                _ => return Err(format!(
                    "unknown crate type: `{part}`, expected one of: {display}",
-                     display = SUPPORTED_CRATE_TYPES.join(", ")
+                    display = ["lib", "rlib", "staticlib", "dylib", "cdylib", "bin", "proc-macro"].join(", ")
                )),
```

Reply:
Ok yeah, let's just do the simplest thing here, since the crate types we accept isn't going to change very often.

Inlining will do, don't bother with the helper methods or the const array.

```rs
pub fn parse_crate_types_from_list(list_list: Vec<String>) -> Result<Vec<CrateType>, String> {
    // NOTE: the param is called `list_list` as we support invocations like:
    //
    // ```
    // --crate-type=rlib,staticlib --crate-type=dylib`
    // ```

    let mut crate_types: Vec<CrateType> = Vec::new();
    for unparsed_crate_type in &list_list {
        for part in unparsed_crate_type.split(',') {
            let new_part = match part {
                "lib" => default_lib_output(),
                "rlib" => CrateType::Rlib,
                "staticlib" => CrateType::Staticlib,
                "dylib" => CrateType::Dylib,
                "cdylib" => CrateType::Cdylib,
                "bin" => CrateType::Executable,
                "proc-macro" => CrateType::ProcMacro,
                _ => {
                    return Err(format!(
                        "unknown crate type: `{part}`, expected one of: {display}",
                        display =
                            ["lib", "rlib", "staticlib", "dylib", "cdylib", "bin", "proc-macro"]
                                .join(", ")
                    ));
                }
            };
            if !crate_types.contains(&new_part) {
                crate_types.push(new_part)
            }
        }
    }

    Ok(crate_types)
}

```